src/main/scala/org/bitcoins/core/protocol/script/ScriptPubKey.scala

```diff
@@ -506,53 +461,179 @@ object CSVScriptPubKey extends Factory[CSVScriptPubKey] {
 
 sealed trait NonStandardScriptPubKey extends ScriptPubKey
 
-object NonStandardScriptPubKey extends Factory[NonStandardScriptPubKey] {
+object NonStandardScriptPubKey extends ScriptFactory[NonStandardScriptPubKey] {
   private case class NonStandardScriptPubKeyImpl(hex : String) extends NonStandardScriptPubKey
 
   override def fromBytes(bytes: Seq[Byte]): NonStandardScriptPubKey = {
-    val asm = ScriptParser.fromBytes(bytes)
+    val asm = RawScriptPubKeyParser.read(bytes).asm
     NonStandardScriptPubKey(asm)
   }
 
   def fromAsm(asm: Seq[ScriptToken]): NonStandardScriptPubKey = {
-    val hex = asm.map(_.hex).mkString
-    NonStandardScriptPubKeyImpl(hex)
+    //everything can be a NonStandardScriptPubkey, thus the trivially true function
+    buildScript(asm,NonStandardScriptPubKeyImpl(_), {_ => true }, "")
   }
 
   def apply(asm : Seq[ScriptToken]) : NonStandardScriptPubKey = fromAsm(asm)
 }
 
-/**
- * Represents the empty ScriptPubKey
- */
+/** Represents the empty ScriptPubKey */
 case object EmptyScriptPubKey extends ScriptPubKey {
-  def hex = ""
+  def hex = "00"
 }
 
-/**
-  * Factory companion object used to create ScriptPubKey objects
-  */
+/** Factory companion object used to create ScriptPubKey objects */
 object ScriptPubKey extends Factory[ScriptPubKey] with BitcoinSLogger {
-  def empty : ScriptPubKey = fromAsm(List())
+  def empty : ScriptPubKey = fromAsm(Nil)
 
-  /**
-    * Creates a scriptPubKey from its asm representation
-    *
-    * @param asm
-    * @return
-    */
+  /** Creates a scriptPubKey from its asm representation */
   def fromAsm(asm : Seq[ScriptToken]) : ScriptPubKey = asm match {
-    case Seq() => EmptyScriptPubKey
+    case Nil => EmptyScriptPubKey
     case _ if P2PKHScriptPubKey.isP2PKHScriptPubKey(asm) => P2PKHScriptPubKey(asm)
     case _ if P2SHScriptPubKey.isP2SHScriptPubKey(asm) => P2SHScriptPubKey(asm)
     case _ if P2PKScriptPubKey.isP2PKScriptPubKey(asm) => P2PKScriptPubKey(asm)
     case _ if MultiSignatureScriptPubKey.isMultiSignatureScriptPubKey(asm) => MultiSignatureScriptPubKey(asm)
     case _ if CLTVScriptPubKey.isCLTVScriptPubKey(asm) => CLTVScriptPubKey(asm)
     case _ if CSVScriptPubKey.isCSVScriptPubKey(asm) => CSVScriptPubKey(asm)
+    case _ if WitnessScriptPubKey.isWitnessScriptPubKey(asm) => WitnessScriptPubKey(asm).get
+    case _ if WitnessCommitment.isWitnessCommitment(asm) => WitnessCommitment(asm)
     case _ => NonStandardScriptPubKey(asm)
   }
 
   def fromBytes(bytes : Seq[Byte]) : ScriptPubKey = RawScriptPubKeyParser.read(bytes)
 
   def apply(asm : Seq[ScriptToken]) : ScriptPubKey = fromAsm(asm)
-}
\ No newline at end of file
+}
+
+/** This type represents a [[ScriptPubKey]] to evaluate a [[ScriptWitness]] */
+sealed trait WitnessScriptPubKey extends ScriptPubKey {
+  def witnessProgram: Seq[ScriptToken]
+  def witnessVersion = WitnessVersion(asm.head.toLong)
```

Comment:
Pretty sure `asm.head.toLong` is wrong, this will work for version programs since `OP_0` is the zero byte, but won't work for any others. For instance when we get a version 1 program, that will be `OP_1` for `witnessVersion`, which is technically `0x51`